JuliaStream.jl/src/ThreadedStream.jl

```diff
@@ -63,12 +63,46 @@ function nstream!(data::VectorData{T}, _) where {T}
   end
 end
 
+# Threads.@threads/Threads.@spawn doesn't support OpenMP's firstprivate, etc
+function static_par_ranged(f::Function, range::Int, n::Int)
+  stride = range ÷ n
+  rem = range % n
+  strides = map(0:n) do i
+    width = stride + (i < rem ? 1 : 0)
+    offset = i < rem ? (stride + 1) * i : ((stride + 1) * rem) + (stride * (i - rem))
+    (offset, width)
+  end
+  ccall(:jl_enter_threaded_region, Cvoid, ())
+  try
+    foreach(wait, map(1:n) do group
+      (offset, size) = strides[group]
+      task = Task(() -> f(group, offset+1, offset+size))
+      task.sticky = true
+      ccall(:jl_set_task_tid, Cvoid, (Any, Cint), task, group-1) # ccall, so 0-based for group
+      schedule(task)
+    end)
+  finally
+    ccall(:jl_exit_threaded_region, Cvoid, ())
+  end
+end
+
 function dot(data::VectorData{T}, _) where {T}
-  partial = zeros(T, Threads.nthreads())
-  Threads.@threads for i = 1:data.size
-    @inbounds partial[Threads.threadid()] += data.a[i] * data.b[i]
+  partial = Vector{T}(undef,  Threads.nthreads())
+  static_par_ranged(data.size, Threads.nthreads()) do group, startidx, endidx
+    acc = zero(T)
+    @fastmath for i = startidx:endidx
```

Comment:
Use `@simd` instead of `@fastmath` that makes sure that only the reduction variable get's reordered

Reply:
Thanks! I didn't really want to use `@fastmath` in a benchmark like STREAM and I'm happy report I get the same runtime with `@simd` on both ThunderX2 and A64fx!

Reply:
On the A64FX node of Isambard, I've got a good speed up with `@simd` when setting the environment variable `JULIA_LLVM_ARGS="-aarch64-sve-vector-bits-min=512"`: https://github.com/JuliaLang/julia/issues/40308#issuecomment-901478623. Note that this requires llvm 12, so Julia v1.7

Reply:
Yeah `@simd` finds the reduction chain, and then adds a subset of the fast math flags onto that chain.

For the A64fx you might also want to set `JULIA_LLVM_ARGS="-aarch64-sve-vector-bits-min=512"` see https://github.com/JuliaLang/julia/issues/40308#issuecomment-901478623

Reply:
BTW, I didn't show in that comment that there is also a good scaling with the size of the floating type (including half-precision `Float16`):
```julia
$ JULIA_LLVM_ARGS="-aarch64-sve-vector-bits-min=512" julia -q
julia> using BenchmarkTools

julia> function sumsimd(x)
           s = zero(eltype(x))
           @simd for xi in x
               s += xi
           end
           s
       end
sumsimd (generic function with 1 method)

julia> @btime sumsimd(x) setup=(x = randn(Float64, 1_000_000))
  191.912 μs (0 allocations: 0 bytes)
1853.0335322487956

julia> @btime sumsimd(x) setup=(x = randn(Float32, 1_000_000))
  80.330 μs (0 allocations: 0 bytes)
400.9806f0

julia> @btime sumsimd(x) setup=(x = randn(Float16, 1_000_000))
  42.761 μs (0 allocations: 0 bytes)
Float16(1.872e3)
```

Related to this, you can watch this lightning talk at JuliaCon 2021 about speedup on using `Float16` on Isambard: https://www.youtube.com/watch?v=btHfZr2C0GA

Reply:
Note that the speedup reported at JuliaCon requires https://github.com/JuliaLang/julia/issues/40216